jobs/populate_db.py

```diff
@@ -8,174 +13,148 @@
-            If it exists, and does not have a value there or in the Annotations layer, add it to the empty_fields list and move to the next field.
-            If it exists, and has a value, add it to completed_fields and move to the next field.
-
-        Step 3: Check for a value in the Annotations layer.
-            If there is a value, add the field to completed_fields.
-            If there is no value, add the field to empty_fields.
-
-        When all the fields have been processed, pass empty_fields to add_tasks and completed_fields to remove_tasks.
-    """
-    completed_fields = []
-    empty_fields = []
-    tool_name = tool["name"]
-    for field in tool["annotations"]:
-        fields_to_skip = [
-            "replaced_by",
-            "deprecated",
-            "experimental",
-            "developer_docs_url",
-            "user_docs_url",
-            "feedback_url",
-            "privacy_policy_url",
-            "for_wikis",
-            "available_ui_languages",
-        ]
-        if field in fields_to_skip:
-            continue
-        elif field in tool:
-            if (tool[field] == [] or tool[field] is None) and (
-                tool["annotations"][field] == [] or tool["annotations"][field] is None
-            ):
-                empty_fields.append(field)
-                continue
-            elif tool[field] != [] or tool[field] is not None:
-                completed_fields.append(field)
-                continue
-        elif tool["annotations"][field] == [] or tool["annotations"][field] is None:
-            empty_fields.append(field)
-        elif tool["annotations"][field] != [] or tool["annotations"][field] is not None:
-            completed_fields.append(field)
-    print({"Empty": empty_fields, "Completed": completed_fields})
-    add_tasks(empty_fields, tool_name)
-    remove_tasks(completed_fields, tool_name)
-
-
-def remove_tasks(fields, tool_name):
-    """
-    Receives fields and a tool name and removes incomplete tasks.
-
-    For each field in the list, the function removes a matching, INCOMPLETE (i.e., one with no value for "user") task from the db, if one exists.
-    """
-    for field in fields:
-        query = text(
-            "DELETE FROM task WHERE field_name = :field_name AND tool_name = :tool AND user IS NULL"
-        ).bindparams(field_name=field, tool=tool_name)
-        db.session.execute(query)
+
+
+def update_tool_table(tools: list[ToolhuntTool], timestamp, **kwargs) -> None:
+    """Upserts tool records and removes stale tools"""
+
+    [upsert_tool(tool, timestamp) for tool in tools]
+
+    remove_stale_tools()
+
+
+def insert_task(tool_name: str, field: str) -> None:
+    """Inserts a tool in the Tool table if it doesn't exist."""
+    query = text(
+        "SELECT * FROM task WHERE field_name = :field AND tool_name = :tool_name"
+    ).bindparams(field=field, tool_name=tool_name)
+    result = db.session.execute(query).all()
+    if len(result) == 0:
```

Comment:
`if not result` would be a better way to write this condition, taking advantage of the fact that empty sequences evaluage to False.

Reply:
Addressed with [c618fa6](https://github.com/wikimedia/toolhunt/pull/56/commits/c618fa62d6053cc13b857fd2a06a46d43c6dafd7)

Reply:
Addressed in [e780c21](https://github.com/wikimedia/toolhunt/pull/56/commits/e780c213ba86602592585388fb407d0d00313e7c)